packages/plugins/experimental/plugin-calls/src/utils/cf-calls-peer.ts

```diff
@@ -0,0 +1,429 @@
+  apiBase: string;
+};
+
+export type Session = {
+  sessionId: string;
+  peerConnection: RTCPeerConnection;
+};
+
+/**
+ * Client for Cloudflare Calls.
+ * API: https://developers.cloudflare.com/calls/https-api/
+ * Inspired by client from https://github.com/threepointone/partyserver/tree/main/packages/partytracks
+ */
+export class CloudflareCallsPeer extends Resource {
+  private readonly _persistentLifecycle = new PersistentLifecycle<Session>({
+    start: () => this._startSession(),
+    stop: (session) => this._stopSession(session),
+  });
+
+  private readonly _config: CloudflareCallsConfig;
+
+  taskScheduler = new FIFOScheduler();
+  pushTrackDispatcher = new BulkRequestDispatcher<
+    {
+      trackName: string;
+      transceiver: RTCRtpTransceiver;
+    },
+    { tracks: TrackObject[] }
+  >(32);
+
+  pullTrackDispatcher = new BulkRequestDispatcher<
+    TrackObject,
+    {
+      trackMap: Map<TrackObject, { resolvedTrack: Promise<MediaStreamTrack>; mid: string }>;
+    }
+  >(32);
+
+  closeTrackDispatcher = new BulkRequestDispatcher(32);
+  constructor(config: CloudflareCallsConfig) {
+    super();
+    this._config = config;
+  }
+
+  protected override async _open(ctx: Context): Promise<void> {
+    await this._persistentLifecycle.open(ctx);
+  }
+
+  protected override async _close(): Promise<void> {
+    await this._persistentLifecycle.close();
+  }
+
+  get session() {
+    return this._persistentLifecycle.state;
+  }
+
+  // TODO(mykola): Expose session errors.
+  get sessionError() {
+    return undefined;
+  }
+
+  private async _startSession() {
+    const peerConnection = await this._createRtcConnection();
+    const sessionId = await this._initiateCallSession();
+    return { sessionId, peerConnection };
+  }
+
+  private async _stopSession(session: Session) {
+    session.peerConnection.close();
+  }
+
+  private async _createRtcConnection(): Promise<RTCPeerConnection> {
+    const peerConnection = new RTCPeerConnection({
+      iceServers: this._config.iceServers,
+      bundlePolicy: 'max-bundle',
+    });
+    peerConnection.addEventListener('connectionstatechange', () => {
+      if (peerConnection.connectionState === 'failed' || peerConnection.connectionState === 'closed') {
+        log.warn(`Calls connection failed with state ${peerConnection.connectionState}`);
+        this._persistentLifecycle.scheduleRestart();
+      }
+    });
+
+    let iceTimeout = -1;
+    peerConnection.addEventListener('iceconnectionstatechange', () => {
+      clearTimeout(iceTimeout);
+      if (peerConnection.iceConnectionState === 'failed' || peerConnection.iceConnectionState === 'closed') {
+        log.warn(`Calls connection to ICEs failed with state ${peerConnection.iceConnectionState}`);
+        this._persistentLifecycle.scheduleRestart();
+      } else if (peerConnection.iceConnectionState === 'disconnected') {
+        // TODO: we should start to inspect the connection stats from here on for
```

Comment:
Is this your TODO or did we inherit this?

Reply:
It is inherited. I would mark them as mine because sooner or later, we need to implement them, and I understand what needs to be done